KomaMRIFiles/src/Sequence/WritePulseq.jl

```diff
@@ -0,0 +1,407 @@
+neccessary to add the input vector `unique_obj_id` which contains the uniques objects (RF,
+Grad, or ADC) with its repective ID.
+"""
+function get_type_blk_obj_id(seq::Sequence, type::String, unique_obj_id::Vector)
+
+    type_blk_obj_id = [[blk, get_obj(s, type), 0] for (blk, s) ∈ enumerate(seq)]
+    for boi ∈ type_blk_obj_id
+        for (obj_unique, id_unique) ∈ unique_obj_id
+            if boi[2] ≈ obj_unique
+                boi[3] = id_unique
+            end
+        end
+    end
+    return type_blk_obj_id
+end
+
+"""
+    output = magsign(v)
+
+Returns the extreme value (positive or negative) which has the maximum absolute value.
+"""
+function magsign(v)
+    a, b = extrema(v)
+    return -a > b ? a : b
+end
+
+"""
+Given an angle in the range [-1, 1], it returns unwrapped values in the range [-0.5, 0.5].
+"""
+function unwrap_angle(ang)
+    (ang <= -0.5 + 1e-7) && return ang + 1.0
+    (0.5 + 1e-7 < ang) && return ang - 1.0
+    return ang
+end
+
+"""
+Given an angle in the range [-0.5, 0.5], it returns unwrapped values in the range [0, 1].
+"""
+function unwrap_negative_angle(ang)
+    return (ang < 0) ? (ang + 1.0) : (ang)
+end
+
+"""
+    rfunique_abs_id, rfunique_ang_id, rfunique_tim_id, id_shape_cnt = get_rfunique(rfunique_obj_id::Vector, id_shape_cnt::Integer, seq::Sequence)
+
+Returns the unique shapes for the magnitude, angle and time of the "rfunique_obj_id" vector.
+Requires an initial integer counter "id_shape_cnt" to asign IDs incrementally.
+"""
+function get_rfunique(rfunique_obj_id::Vector, id_shape_cnt::Integer, seq::Sequence)
+    # Find the unique shapes (magnitude, phase and time shapes) and assign IDs
+    rfunique_abs_id, rfunique_ang_id, rfunique_tim_id = [], [], []
+    for (obj,_) ∈ rfunique_obj_id
+        shape_abs = abs.(obj.A) / maximum(abs.(obj.A))
+        if all([!(length(shape_abs) == length(shape_abs_unique) && shape_abs ≈ shape_abs_unique) for (shape_abs_unique,_) ∈ rfunique_abs_id])
+            push!(rfunique_abs_id, [shape_abs, id_shape_cnt])
+            id_shape_cnt += 1
+        end
+        shape_ang = -angle.(obj.A)/2π  # [-0.5, 0.5]
+        if all([!(length(shape_ang) == length(shape_ang_unique) && all(isapprox.(unwrap_angle.(shape_ang .- shape_ang[1]), unwrap_angle.(shape_ang_unique .- shape_ang_unique[1]), atol=1e-7))) for (shape_ang_unique,_) ∈ rfunique_ang_id])
+            push!(rfunique_ang_id, [unwrap_negative_angle.(shape_ang), id_shape_cnt])   # this always garanties non-negative shaped angles
+            id_shape_cnt += 1
+        end
+        if isa(obj.T, Vector{<:Number})
+            shape_tim = cumsum([0; obj.T]) / seq.DEF["RadiofrequencyRasterTime"]
+            if all([!(length(shape_tim) == length(shape_tim_unique) && shape_tim ≈ shape_tim_unique) for (shape_tim_unique,_) ∈ rfunique_tim_id])
+                push!(rfunique_tim_id, [shape_tim, id_shape_cnt])
+                id_shape_cnt += 1
+            end
+        end
+    end
```

Comment:
All of these comparisons are not defined in isapprox? what is the purpose of them then